store/src/java/com/zimbra/cs/imap/ImapSearch.java

```diff
@@ -42,15 +42,15 @@ static String sequenceAsSearchTerm(ImapFolder i4folder, TreeSet<ImapMessage> i4s
         i4set.remove(null);
         if (i4set.isEmpty()) {
             return "item:none";
-        } else if (abbreviateAll && isAllMessages(i4folder, i4set)) {
+        } else if (abbreviateAll) {
```

Comment:
This looks wrong to me.  It looks like this line is testing whether or not all messages should be returned and if they are, whether an abbreviated search should be used.  This change would cause all messages to be returned, even when only some have been requested.  As Greg said, probably need tests for the problem being addressed but also tests for searching for subsets.

Reply:
The issue is that the test for length isn't used here so if the i4set contains 10000 items they will all be included in the entire search query without this change.

Perhaps we need to rethink all of this?

Reply:
i.e.  abbreviateAll == true && !isAllMesssages()

Reply:
Yes, I think a rethink is needed for the case when isAllMessages() returns false (or even when abbreviateAll is false)  I think you said there is paging code but that gets involved too late.  I suspect the fix will require making the paging effective.